includes/class-wc-gateway-stripe.php

```diff
@@ -1191,4 +1061,204 @@ public function retry_after_error( $response, $order, $retry, $force_save_source
+     */
+    public function create_and_confirm_intent( $order, $prepared_source ) {
+        // The request for a charge contains metadata for the intent.
+        $full_request = $this->generate_payment_request( $order, $prepared_source );
+
+        $request = array(
+            'source'               => $prepared_source->source,
+            'amount'               => WC_Stripe_Helper::get_stripe_amount( $order->get_total() ),
+            'currency'             => get_woocommerce_currency(),
+            'description'          => $full_request['description'],
+            'metadata'             => $full_request['metadata'],
+            'statement_descriptor' => $full_request['statement_descriptor'],
+            'allowed_source_types' => array(
+                'card',
+            ),
+        );
+
+        if ( $prepared_source->customer ) {
+            $request['customer'] = $prepared_source->customer;
+        }
+
+        // Create an intent that awaits an action.
+        $intent = WC_Stripe_API::request( $request, 'payment_intents' );
+        if ( ! empty( $intent->error ) ) {
+            return $intent;
+		}
+
+		$order->add_order_note( sprintf( __( 'Stripe PaymentIntent initiated (ID: %s)', 'woocommerce-gateway-stripe' ), $intent->id ) );
+
+        // Try to confirm the intent & capture the charge (if 3DS is not required).
+        $request = array(
+			'source' => $request['source'],
+        );
+        $confirmed_intent = WC_Stripe_API::request( $request, "payment_intents/$intent->id/confirm" );
+
+        if ( ! empty( $confirmed_intent->error ) ) {
+			return $confirmed_intent;
+        }
+
+        // Save the intent ID to the order
+		$this->save_intent_to_order( $order, $confirmed_intent );
+
+		// Save a note about the status of the intent.
+		if ( 'succeeded' === $confirmed_intent->status ) {
+			$order->add_order_note( sprintf( __( 'Stripe PaymentIntent succeeded (ID: %s)', 'woocommerce-gateway-stripe' ), $intent->id ) );
+		} else if ( 'requires_source_action' === $confirmed_intent->status ) {
+			$order->add_order_note( sprintf( __( 'Stripe PaymentIntent requires authentication (ID: %s)', 'woocommerce-gateway-stripe' ), $intent->id ) );
+		}
+
+        return $confirmed_intent;
+	}
+
+	/**
+	 * Updates an existing intent with updated amount, source, and customer.
+	 *
+	 * @param object   $intent          The existing intent object.
+	 * @param WC_Order $order           The order.
+	 * @param object   $prepared_source Currently selected source.
+	 * @return object                   An updated intent.
+	 */
+	public function update_existing_intent( $intent, $order, $prepared_source ) {
+		$request = array();
+
+		if ( $prepared_source->source !== $intent->source ) {
+			$request['source'] = $prepared_source->source;
+		}
+
+		$new_amount = WC_Stripe_Helper::get_stripe_amount( $order->get_total() );
+		if ( $intent->amount !== $new_amount ) {
```

Comment:
I assume there is no way we can get oddities from comparing floating point numbers here?

Reply:
There should be no floating point numbers here. `get_stripe_amount` and Stripe's API work with positive integers.